CODING_GUIDELINES.md

```diff
@@ -6,30 +6,82 @@ This document contains the coding guidelines for the ASPIRE Python project.
-and [PEP 257 Docstring Conventions](https://www.python.org/dev/peps/pep-0257/) guidelines. [This webpage](https://realpython.com/documenting-python-code/) is a good guide for documenting Python code. 
-For automatic checking of style problems, [Pycharm from Jet Brains](https://www.jetbrains.com/pycharm/) 
-is a good candidate which has the professional version for academic users with free license fee. 
-
-
-## Best Practices
-
-For beginners in scientific computing, some general guides are suggested in Greg Wilson's papers as below:
-
- 1. [Good Enough Practices in Scientific Computing]( https://doi.org/10.1371/journal.pcbi.1005510) 
- 2. [Best Practices for Scientific Computing]( https://doi.org/10.1371/journal.pbio.1001745) 
+and [PEP 257 Docstring Conventions](https://www.python.org/dev/peps/pep-0257/) guidelines. [This webpage](https://realpython.com/documenting-python-code/) is a good guide for documenting Python code.
+For an editor with some automatic checking of style problems, [Pycharm from Jet Brains](https://www.jetbrains.com/pycharm/)
+is a good candidate which has the professional version for academic users with free license fee.
 
+In the future compoenents of PEP8 and PEP257 will be checked programmatically.  Documentation of a standard tool to perform
+these operations locally will be provided to developers at that time.
 
 ## Source Code Control
 
-As an open source software, we use Git and GitHub for our source code control. 
-The following links are useful tutorials. 
+As an open source software, we use Git with GitHub for our source code control.
+The following links are useful tutorials.
 1. Git tutorials from [Atalassian](https://www.atlassian.com/git/tutorials) and [Git-SCM](https://git-scm.com/docs/gittutorial)
-2. GitHub guides from [GitHub](https://guides.github.com/) 
+2. GitHub guides from [GitHub](https://guides.github.com/)
 
-## Git Workflow Changes
+## Git Workflow
 
-Our project follows [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).
+Our project follows a reduced [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).
 Please submit any PRs against the `develop` branch.
 
 ![Gitflow Diagram](https://wac-cdn.atlassian.com/dam/jcr:61ccc620-5249-4338-be66-94d563f2843c/05%20(2).svg?cdnVersion=357)
 
+The basic idea is that we have `feature` branches,  `develop` as an integration focused branch, and `master` for releases.
+We do not currently have a need for a dedicated release or staging branch; in our case `develop` also serves this purpose.
+
+Developers are welcome to maintain their own fork, but this is not required.
+
+When `develop` is at a milestone where we want to cut a release we will commit with a PEP440 compliant `Major.Minor` version
+scheme, tag it similarly, and complete a merge to master following the controls below. It is at this point the
```

Comment:
Since we do not have formal release yet, our previous experience is that we will keep master updated when we update the develop branch. In this way, the users and the researchers in the group cloning the master branch will know the updates and resolve possible conflicts as soon as possible. This is based on our current situation in a stage of rapid changing. Once the first release is done, we think the master branch will be updated less frequently.

Reply:
ASPIRE already has (public) releases, they just haven't been kept up to date in the last year... The procedure does not appear to have been followed during this time. We talked about this today in our meeting.

(Politely) Your statement about conflicts doesn't make sense to me. (Calmly) Code conflicts happen when developers are not aware and coordinating their efforts, staying up to date, or reviewing each other's code if they are upstream.  People can clone any branch they want for their research, that is their choice and responsibility. Generally, you probably want them working off a release, or coordinating with a developer if they are going to be on an active working branch.  This is just a guideline and I am open to change, but we shouldn't say we're using Git flow or whatever, and then just do whatever we want when it is convenient or there is a gap in oversight. 

Perhaps we're thinking different timescales or something. We can discuss more, I'm very curious why you think good practices won't work for us as a team. I have seen it bring a lot of value to small groups of talented people in particular.

Reply:
I tried to make clean when we should update master since this will be very different to our previous practice. Probably include Joakim's suggestions too.

Reply:
Sure, we should be able to go through this with him later this week.  I don't recall anything he suggested that is missing. Do you have something in mind?

Reply:
So there's a few things here:

- I agree that we should be conservative in merging `develop` into `master`. We want things to reach a relatively stable state before doing so. Otherwise, what is the point of keeping two separate branches (as opposed to just merge everything into `master` as we add features)?
- Regarding users. Right now, we don't have too many of them. For the ones we do have, it would be good to provide a relatively stable experience.
- For developers, we expect them to work off of `develop` and merge their changes in there. As Garrett says, this needs to be coordinated with other developers as much as possible. In that sense, I don't think there'll be a problem “keeping up” with updates and potential conflicts.

Also, `master` should be in backticks.

Reply:
backticked master.

Reply:
OK. So we agree to update master less frequently. Probably the code owner should perform the merge from the develop to the master?  By the way we are trying to use the terms from agile development framework. But a lot rules might be not applicable for academic environment. I also suggest adding the PI to the code owner list so that we will not surprise academic users.

Reply:
> Probably the code owner should perform the merge from the develop to the master?

The issue is that `develop` will contain changes from all over the code, so there won't be a single code owner. It'll probably just make sense for me to do this periodically when we feel the code is stable enough.

> By the way we are trying to use the terms from agile development framework.

I'm not so familiar with this. Can you elaborate?

Reply:
I got some ideas recently but might be not correctly. An agile develop team includes the product owner (code owner) for laying out needs and goals , the scrum master for taking charging of all aspects of code development, and a few developers working on different code components. The team will work on a project within a couple of weeks and meet every day for daily updates. I think your role is closer to both product owner and scrum master. But clearly we can not fix the whole organization and timeline.

Reply:
product owner and code owner are not the same thing. They're not related.

As someone who was certified in a few different flavors of that PM stuff, I have to say min agile team is like 5 people, minimum, and they all need to be off a certain software engineering caliber or it doesn't work at all. (To treat developers as widgets they must be interchangeable, that is not the case here, not even close).

Another problem with agile development in our setting is we have no actual recourse for problems or rework, besides increasing the meeting or collaboration burden. This probably won't have the desired effects.

Otherwise, yeah, use the terms away, I'm totally fine with them. :popcorn: